diff -up a/dwm.c b/dwm.c
--- a/dwm.c	2015-11-12 17:52:36.645173774 +0300
+++ b/dwm.c	2015-11-12 18:23:43.481155570 +0300
@@ -210,6 +210,7 @@ static void resizeclient(Client *c, int
 static void resizemouse(const Arg *arg);
 static void restack(Monitor *m);
 static void run(void);
+static void runautostart(void);
 static void scan(void);
 static int sendevent(Client *c, Atom proto);
 static void sendmon(Client *c, Monitor *m);
@@ -1710,6 +1711,13 @@ run(void)
 }
 
 void
+runautostart(void)
+{
+    system("cd ~/.dwm; ./autostart_blocking.sh");
+    system("cd ~/.dwm; ./autostart.sh&");
+}
+
+void
 scan(void)
 {
 	unsigned int i, num;
@@ -2514,6 +2522,7 @@ main(int argc, char *argv[])
 	checkotherwm();
 	setup();
 	scan();
+    runautostart();
 	run();
 	cleanup();
 	XCloseDisplay(dpy);
